### PR TITLE
Refactor host and add managed fields

### PR DESCRIPTION
**Pull request by MrFreezeex**
The first commit add a `managed` field to host. This allow to not specify the mac address/ip address if you set it to false (not sure if this is redundant to manage_build, I think it should be ok like that though).

The second commit refactor a bit the ForemanHost mainly adding a bunch of omitempty (and having *int instead of int so that omitempty actually works), removing the custom marshaling and making `ComputeAttributes` type a little more precise (from interface{} to map[string]interface{}). I also removed all the `GetOk` call while building the ForemanHost struct as if the schema attribute is `computed` this method always return ~~true~~ false (https://github.com/hashicorp/terraform-plugin-sdk/blob/9bdb216a4ab835003a624b0a15ac39c58ef2861e/helper/schema/resource_diff.go#L386).

### CODE REVIEW

**Comment by MrFreezeex** [comment]
Ah yes indeed, I need to fix some type in the tests

**Comment by lhw** [comment]
>  I also removed all the GetOk call while building the ForemanHost struct as if the schema attribute is computed this method always return true (https://github.com/hashicorp/terraform-plugin-sdk/blob/9bdb216a4ab835003a624b0a15ac39c58ef2861e/helper/schema/resource_diff.go#L386).
Wow. I never realized that. Thats all still code from wayfair and I presumed its basically just a != nil check.

**Comment by MrFreezeex** [comment]
> > I also removed all the GetOk call while building the ForemanHost struct as if the schema attribute is computed this method always return true (https://github.com/hashicorp/terraform-plugin-sdk/blob/9bdb216a4ab835003a624b0a15ac39c58ef2861e/helper/schema/resource_diff.go#L386).
> 
> Wow. I never realized that. Thats all still code from wayfair and I presumed its basically just a != nil check.

Yes... Fortunately for us the foreman api is forgiving if you give him a bunch of id that equals to 0 etc so that sort of works.

**Comment by MrFreezeex** [comment]
> Ah yes indeed, I need to fix some type in the tests

Should be good now :+1:

File: foreman/resource_foreman_host.go [hunk 1]
```diff
@@ -95,6 +97,13 @@ func resourceForemanHost() *schema.Resource {
 					"boot to PXE and power on. Defaults to `false`.",
 			},
 
+			"managed": &schema.Schema{
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     true,
+				Description: "Whether or not this host is managed by Foreman.",
+			},
+
```
Comment (lhw) [inline foreman/resource_foreman_host.go:106]: Is this essentially the same as the `manage_build` parameter below?

File: foreman/api/host.go [hunk 2]
```diff
@@ -154,129 +155,6 @@ type BMCBoot struct {
...
-
-	if _, ok = fhMap["parameters"]; ok {
-		for _, parameter := range fhMap["parameters"].([]interface{}) {
-			param := parameter.(map[string]interface{})
-			fh.HostParameters = append(fh.HostParameters, ForemanKVParameter{
-				Name:  param["name"].(string),
-				Value: param["value"].(string),
-			})
-		}
-	}
```
Comment (lhw) [inline foreman/api/host.go:258]: I feel like this might break some compatibility. I will have to test this.

File: foreman/resource_foreman_host.go (hunk 1 above)
Comment (MrFreezeex) [reply foreman/resource_foreman_host.go:106]: Well `manage_build` have some kind of logic to do more calls related to power and so on. This is is a direct mapping to the `managed` field of foreman host and that's about it. But maybe it could be merged somehow with manage_build but I didn't wanted to break existing behaviors so I don't really know...

File: foreman/api/host.go [hunk 3]
```diff
@@ -154,129 +155,6 @@ type BMCBoot struct {
...
-	fh.MediumId = unmarshalInteger(fhMap["medium_id"])
-	fh.ImageId = unmarshalInteger(fhMap["image_id"])
-	fh.ModelId = unmarshalInteger(fhMap["model_id"])
-	fh.ComputeResourceId = unmarshalInteger(fhMap["compute_resource_id"])
-	fh.ComputeProfileId = unmarshalInteger(fhMap["compute_profile_id"])
-
-	// Foreman returns FQDN as Name but doesnt accept it as Name in return. Great times
-	if fh.DomainName != "" && strings.Contains(fh.ForemanObject.Name, fh.DomainName) {
-		fh.ForemanObject.Name = strings.Replace(fh.ForemanObject.Name, "."+fh.DomainName, "", 1)
-	}
```
Comment (agriffit79) [inline foreman/api/host.go:275]: With this code removed terraform wants to replace all my hosts as the change in `name` forces replacement.

Comment (MrFreezeex) [reply foreman/api/host.go:275]: Oh indeed now that you mention it I think I had also the issue but didn't realize, I think the correct way to fix it would be a supress diff on the name. Will take a look, thanks!

Comment (MrFreezeex) [reply foreman/api/host.go:275]: I added a new commit to fix this and seems to work for me. Did it fix your issue?

Comment (agriffit79) [reply foreman/api/host.go:275]: That's resolved, but I'm seeing other diffs now with `interface_atttributes` and `puppet_class_ids`. Need to dig into it some more but don't have time right now. Should have some time on Friday.

Comment (MrFreezeex) [reply foreman/api/host.go:275]: Oh thanks reading the code I might know why: it's the two parameters being parsed with differents keys in `foremanHostJSON`. Maybe the API has different keys depending if you create the object or read it... I will test things and try to find a fix for this :+1:.

Comment (MrFreezeex) [reply foreman/api/host.go:275]: I pushed a new fix for that as well please test it whenever you have some time!

File: foreman/api/host.go (hunk 2 above)
Comment (lhw) [reply foreman/api/host.go:258]: This didn't break anything in our infrastructure but I will test a bit more with some corner cases.

File: foreman/resource_foreman_host.go (hunk 1 above)
Comment (lhw) [reply foreman/resource_foreman_host.go:114]: I think they should either be merged if the logic permits it or otherwise the keys need to be aptly named so that the end user is not confused when reading the documentation.

Comment (MrFreezeex) [reply foreman/resource_foreman_host.go:114]: Also BTW It's this field probably that was probably asked here https://github.com/HanseMerkur/terraform-provider-foreman/issues/52

File: foreman/api/host.go (hunk 3 above)
Comment (agriffit79) [reply foreman/api/host.go:275]: Yes, looks ok now with my simple test stack.

File: foreman/resource_foreman_host.go (hunk 1 above)
Comment (agriffit79) [reply foreman/resource_foreman_host.go:114]: The reason I added the `manage_build` param was to prevent the provider trying to set power states when creating hosts. It could be this param could be removed and the functionality merged with the more generic `managed` param?

Comment (agriffit79) [reply foreman/resource_foreman_host.go]: The only issue is this is effectively a breaking change, as we are removing an existing attribute. Not sure what the best practice way is for handling that? Although the attribute has only been present for a few weeks, so maybe no is using it yet...

Comment (MrFreezeex) [reply foreman/resource_foreman_host.go]: We could handle both variables for a given time with a deprecation of `manage_build`, but since it has just been released we could probably just remove `managed_build` like I did and announce the breaking change in the release notes. Happy to do the change to also handle `managed build` with a deprecation notice if you find it necessary though!

Comment (lhw) [reply foreman/resource_foreman_host.go]: My suggestion would be to keep the attribute with with the option `Removed` and increase the version number to 0.5.0
```
diff --git a/foreman/resource_foreman_host.go b/foreman/resource_foreman_host.go
index 882080f..a649e43 100644
--- a/foreman/resource_foreman_host.go
+++ b/foreman/resource_foreman_host.go
@@ -104,7 +104,13 @@ func resourceForemanHost() *schema.Resource {
 					"calls, having this enabled will force a host to poweroff, set next " +
 					"boot to PXE and power on. Defaults to `false`.",
 			},
-
+			"manage_build": &schema.Schema{
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     true,
+				Removed:     "The feature was merged into the new key managed",
+				Description: "Create host only, don't set build status or manage power states",
+			},
 			"managed": &schema.Schema{
 				Type:     schema.TypeBool,
 				Optional: true,
```

Comment (MrFreezeex) [reply foreman/resource_foreman_host.go]: Sure, I just did this. Also added a note in the description to that this appears in the auto generated doc somehow.

**Review by lhw — APPROVED** [review]
lgtm
